go/fix-chain/fixer.go

```diff
@@ -0,0 +1,282 @@
+		}
+		if ci[n] > cj[n] {
+			return false
+		}
+	}
+	return false
+}
+func (b Bag) Swap(i, j int) {
+	t := b.certs[i]
+	b.certs[i] = b.certs[j]
+	b.certs[j] = t
+}
+
+func HashBag(bag []*x509.Certificate) [HashSize]byte {
+	b := Bag{certs: bag}
+	sort.Sort(b)
+	return HashChain(b.certs)
+}
+
+func HexHash(s *x509.Certificate) string {
+	h := Hash(s)
+	return hex.EncodeToString(h[:])
+}
+
+func dumpChain(name string, certs []*x509.Certificate) {
+	for i, cert := range certs {
+		log.Printf("%s %d: %s %s", name, i, HexHash(cert),
+			cert.Subject.CommonName)
+	}
+}
+
+func dumpChains(name string, chains [][]*x509.Certificate) {
+	for i, chain := range chains {
+		n := fmt.Sprintf("%s %d", name, i)
+		dumpChain(n, chain)
+	}
+}
+
+type DedupedChain struct {
+	certs []*x509.Certificate
+}
+
+func (d *DedupedChain) AddCert(cert *x509.Certificate) {
+	// Check that the certificate isn't being added twice.
+	for _, c := range d.certs {
+		if c.Equal(cert) {
+			return
+		}
+	}
+	d.certs = append(d.certs, cert)
+}
+
+func (d *DedupedChain) Dump(name string) {
+	dumpChain(name, d.certs)
+}
+
+type Fix struct {
+	cert  *x509.Certificate
+	chain *DedupedChain
+	opts  *x509.VerifyOptions
+	fixer *Fixer
+}
+
+// Returns a partially filled FixError on error
+func augmentIntermediates(pool *x509.CertPool, url string,
+	u *URLCache) *FixError {
+	r := urlReplacement(url)
+	if r != nil {
+		log.Printf("Replaced %s: %+v", url, r)
+		for _, c := range r {
+			pool.AddCert(c)
+		}
+		return nil
+	}
+	body, err := u.getURL(url)
+	if err != nil {
+		//log.Printf("can't get URL body from %s: %s", url, err)
+		return &FixError{Type: CannotFetchURL, URL: url, Error: err}
+	}
+	//log.Print(body)
+	icert, err := x509.ParseCertificate(body)
+	if err != nil {
+		s, _ := pem.Decode(body)
+		if s != nil {
+			icert, err = x509.ParseCertificate(s.Bytes)
+		}
+	}
+
+	if err != nil {
+		//log.Fatalf("failed to parse certificate from %s: %s", url, err)
+		return &FixError{Type: ParseFailure, URL: url, Bad: body,
+			Error: err}
+	}
+	pool.AddCert(icert)
+	return nil
+}
+
+func fixChain(fix *Fix) {
+	d2 := *fix.chain
+	d2.AddCert(fix.cert)
+	for _, c := range d2.certs {
+		urls := c.IssuingCertificateURL
+		for _, url := range urls {
+			//log.Printf("fetch issuer %d from %s", i, url)
+			ferr := augmentIntermediates(fix.opts.Intermediates,
+				url, fix.fixer.cache)
+			if ferr != nil {
+				ferr.Cert = fix.cert
+				ferr.Chain = fix.chain.certs
+				fix.fixer.errors <- ferr
+			}
+			chain, err := fix.cert.Verify(*fix.opts)
+			if err == nil {
+				//dumpChains("fixed", chain)
+				fix.fixer.fixed++
+				fix.fixer.log.PostChains(chain)
+				return
+			}
+		}
+	}
+	//log.Printf("failed to fix certificate for %s", fix.cert.Subject.CommonName)
+	fix.fixer.notfixed++
+	fix.fixer.errors <- &FixError{Type: FixFailed, Cert: fix.cert,
+		Chain: fix.chain.certs}
+}
+
+type Fixer struct {
+	fix    chan *Fix
+	active uint32
+	// Counters may not be entirely accurate due to non-atomicity
```

Comment:
I do on counters that need to be accurate. Figured we didn't need the overhead on these, but could.